pkg/engine/plugins.go

```diff
@@ -184,25 +183,31 @@ func gatherPluginsFromSnapshot(plugctx *plugin.Context, target *deploy.Target) (
 	return set, nil
 }
 
-// ensurePluginsAreInstalled inspects all plugins in the plugin set and, if any plugins are not currently installed,
+// EnsurePluginsAreInstalled inspects all plugins in the plugin set and, if any plugins are not currently installed,
 // uses the given backend client to install them. Installations are processed in parallel, though
-// ensurePluginsAreInstalled does not return until all installations are completed.
-func ensurePluginsAreInstalled(ctx context.Context, d diag.Sink,
-	plugins pluginSet, projectPlugins []workspace.ProjectPlugin,
-) error {
-	logging.V(preparePluginLog).Infof("ensurePluginsAreInstalled(): beginning")
+// EnsurePluginsAreInstalled does not return until all installations are completed.
+func EnsurePluginsAreInstalled(ctx context.Context, plugins PluginSet, reinstall bool) error {
+	logging.V(preparePluginLog).Infof("EnsurePluginsAreInstalled(): beginning")
 	var installTasks errgroup.Group
 	for _, plug := range plugins.Values() {
 		if plug.Name == "pulumi" && plug.Kind == apitype.ResourcePlugin {
-			logging.V(preparePluginLog).Infof("ensurePluginsAreInstalled(): pulumi is a builtin plugin")
+			logging.V(preparePluginLog).Infof("EnsurePluginsAreInstalled(): pulumi is a builtin plugin")
 			continue
 		}
 
-		path, err := workspace.GetPluginPath(d, plug.Kind, plug.Name, plug.Version, projectPlugins)
-		if err == nil && path != "" {
-			logging.V(preparePluginLog).Infof(
-				"ensurePluginsAreInstalled(): plugin %s %s already installed", plug.Name, plug.Version)
-			continue
+		if !reinstall {
```

Comment:
Nit: Worth keeping the comment?

```suggestion
		// If the plugin already exists, don't download it unless `reinstall` was specified.
		if !reinstall {
```

Reply:
can't hurt